pdns/ws-auth.cc

```diff
@@ -2270,6 +2270,14 @@ static void apiServerCacheFlush(HttpRequest* req, HttpResponse* resp) {
 
   DNSName canon = apiNameToDNSName(req->getvars["domain"]);
 
+  SOAData sd;
+  UeberBackend B;
+  if (g_zoneCache.isEnabled() && B.getSOAUncached(canon, sd)) {
+    // zone exists (uncached), add/update it in the zone cache.
+    // Handle this first, to avoid concurrent queries re-populating the other caches.
+    g_zoneCache.add(canon, sd.domain_id);
+  }
```

Comment:
```suggestion
  int zoneId{0};
  DomainInfo di;
  if (g_zoneCache.isEnabled() && !g_zoneCache.getEntry(canon, zoneId)) {
    UeberBackend B;
    if (B.getDomainInfo(canon, di, false)) {
      // zone exists (uncached), add/update it in the zone cache.
      // Handle this first, to avoid concurrent queries re-populating the other caches.
      g_zoneCache.add(canon, di.id);
    }
  }
```
Prevent some DB actions for zones already present in the zonecache and add zones without SOA records as well.

Reply:
I like it, but I'll point out that changing the zoneId will not work with this code as is.

Reply:
added.

Reply:
I think it is better to remove "&& !g_zoneCache.getEntry(canon, zoneId)". 
We nuke the entire query cache now, so one additional query will probably not hurt.

Reply:
Agreed, done.